parser-typechecker/src/Unison/Lexer.hs

```diff
@@ -882,6 +908,18 @@ lexemes' eof = P.optional space >> do
 
     close = close' Nothing
 
+    blockDelimiter :: [String] -> P String -> P [Token Lexeme]
+    blockDelimiter open closeP = do
+      (pos1, close, pos2) <- positioned $ closeP
+      env                 <- S.get
+      case findClose open (layout env) of
+        Nothing -> err pos1 (UnexpectedDelimiter (quote close))
+          where quote s = "'" <> s <> "'"
+        Just (block, n) -> do
+          S.put (env { layout = drop n (layout env), opening = Just block })
+          let opens = [Token (Reserved close) pos1 pos2]
+          pure $ replicate (n-1) (Token Close pos1 pos2) ++ opens
```

Comment:
The `n-1` here is key. Close as many blocks as necessary to get us back to the bracket block.